File: src/main/java/org/radarbase/fcm/dto/FcmNotificationDto.java
/*
 *
 *  *
 *  *  * Copyright 2018 King's College London
 *  *  *
 *  *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  *  * you may not use this file except in compliance with the License.
 *  *  * You may obtain a copy of the License at
 *  *  *
 *  *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *  *
 *  *  * Unless required by applicable law or agreed to in writing, software
 *  *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  * See the License for the specific language governing permissions and
 *  *  * limitations under the License.
 *  *  *
 *  *
 *
 */

package org.radarbase.fcm.dto;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;
import org.radarbase.appserver.entity.Notification;
import org.springframework.format.annotation.DateTimeFormat;

import javax.validation.constraints.NotNull;
import java.io.Serializable;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.Objects;

/**
 * @author yatharthranjan
 */
@Getter
@EqualsAndHashCode
@ToString
public class FcmNotificationDto implements Serializable {

    private static final long serialVersionUID = 3L;

    private Long id;

    @NotNull
    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
    private LocalDateTime scheduledTime;

    private boolean delivered;

    @NotNull
    private String title;

    private String body;

    private int ttlSeconds;

    @NotNull
    private String sourceId;

    private String fcmMessageId;

    @NotNull
    private String type;

    @NotNull
    private String appPackage;

    @NotNull
    private String sourceType;

    public FcmNotificationDto(Notification notificationEntity) {
        this.id = notificationEntity.getId();
        this.scheduledTime = LocalDateTime.ofInstant(notificationEntity.getScheduledTime(), ZoneOffset.UTC);
        this.title = notificationEntity.getTitle();
        this.body = notificationEntity.getBody();
        this.delivered = notificationEntity.isDelivered();
        this.fcmMessageId = notificationEntity.getFcmMessageId();
        this.sourceId = notificationEntity.getSourceId();
        this.type = notificationEntity.getType();
        this.appPackage = notificationEntity.getAppPackage();
        this.sourceType = notificationEntity.getSourceType();
        this.ttlSeconds = notificationEntity.getTtlSeconds();
    }

    public FcmNotificationDto() {

    }

    public FcmNotificationDto setId(Long id) {
        this.id = id;
        return this;
    }

    public FcmNotificationDto setScheduledTime(LocalDateTime scheduledTime) {
        this.scheduledTime = scheduledTime;
        return this;
    }

    public FcmNotificationDto setDelivered(boolean delivered) {
        this.delivered = delivered;
        return this;
    }

    public FcmNotificationDto setTitle(String title) {
        this.title = title;
        return this;
    }

    public FcmNotificationDto setBody(String body) {
        this.body = body;
        return this;
    }

    public FcmNotificationDto setTtlSeconds(int ttlSeconds) {
        this.ttlSeconds = ttlSeconds;
        return this;
    }

    public FcmNotificationDto setFcmMessageId(String fcmMessageId) {
        this.fcmMessageId = fcmMessageId;
        return this;
    }

    public FcmNotificationDto setSourceId(String sourceId) {
        this.sourceId = sourceId;
        return this;
    }

    public FcmNotificationDto setType(String type) {
        this.type = type;
        return this;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof FcmNotificationDto)) return false;
        FcmNotificationDto that = (FcmNotificationDto) o;
        return delivered == that.delivered &&
                ttlSeconds == that.ttlSeconds &&
                Objects.equals(scheduledTime, that.scheduledTime) &&
                Objects.equals(title, that.title) &&
                Objects.equals(body, that.body) &&
                Objects.equals(type, that.type) &&
                Objects.equals(appPackage, that.appPackage) &&
                Objects.equals(sourceType, that.sourceType);
    }

    public String getAppPackage() {
        return appPackage;
    }

    public FcmNotificationDto setAppPackage(String appPackage) {
        this.appPackage = appPackage;
        return this;
    }

    public String getSourceType() {
        return sourceType;
    }

    public FcmNotificationDto setSourceType(String sourceType) {
        this.sourceType = sourceType;
        return this;
    }

    @Override
    public int hashCode() {
        return Objects.hash(scheduledTime, delivered, title, body, ttlSeconds, type, appPackage, sourceType);
    }
}

File: microservices/core/src/main/kotlin/org/radarbase/appserver/microservices/core/dto/fcm/FcmNotificationDto.kt
/*
 * Copyright 2025 King's College London
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarbase.appserver.microservices.core.dto.fcm

import com.fasterxml.jackson.annotation.JsonFormat
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.Size
import kotlinx.serialization.Serializable
import org.radarbase.appserver.microservices.core.serialization.InstantSerializer
import java.time.Instant
import org.radarbase.appserver.microservices.core.entity.Notification
import org.radarbase.appserver.microservices.core.serialization.InstantSerializer
import org.radarbase.appserver.microservices.core.utils.equalTo
import java.util.Objects

@Serializable
class FcmNotificationDto(
    var id: Long? = null,

    @field:NotNull
    @Serializable(with = InstantSerializer::class)
    var scheduledTime: Instant? = null,

    var delivered: Boolean = false,

    @field:NotEmpty
    var title: String? = null,

    var body: String? = null,

    var ttlSeconds: Int = 0,

    @field:NotEmpty
    var sourceId: String? = null,

    var fcmMessageId: String? = null,

    var fcmTopic: String? = null,

    var fcmCondition: String? = null,

    @field:NotEmpty
    var type: String? = null,

    @field:NotEmpty
    var appPackage: String? = null,

    @field:NotEmpty
    var sourceType: String? = null,

    @field:Size(max = 100)
    var additionalData: Map<String?, String?>? = null,

    var priority: String? = null,

    var sound: String? = null,

    var badge: String? = null,

    var subtitle: String? = null,

    var icon: String? = null,

    var color: String? = null,

    var bodyLocKey: String? = null,

    var bodyLocArgs: String? = null,

    var titleLocKey: String? = null,

    var titleLocArgs: String? = null,

    var androidChannelId: String? = null,

    var tag: String? = null,

    var clickAction: String? = null,

    var emailEnabled: Boolean = false,

    var emailTitle: String? = null,

    var emailBody: String? = null,

    var mutableContent: Boolean = false,

    @Serializable(with = InstantSerializer::class)
    var createdAt: Instant? = null,

    @Serializable(with = InstantSerializer::class)
    var updatedAt: Instant? = null
) {
    constructor(notification: Notification) : this(
        id = notification.id,
        scheduledTime = notification.scheduledTime,
        delivered = notification.delivered == true,
        title = notification.title,
        body = notification.body,
        ttlSeconds = notification.ttlSeconds ?: 0,
        sourceId = notification.sourceId,
        fcmMessageId = notification.fcmMessageId,
        fcmTopic = notification.fcmTopic,
        fcmCondition = notification.fcmCondition,
        type = notification.type,
        appPackage = notification.appPackage,
        sourceType = notification.sourceType,
        additionalData = notification.additionalData,
        priority = notification.priority,
        sound = notification.sound,
        badge = notification.badge,
        subtitle = notification.subtitle,
        icon = notification.icon,
        color = notification.color,
        bodyLocKey = notification.bodyLocKey,
        bodyLocArgs = notification.bodyLocArgs,
        titleLocKey = notification.titleLocKey,
        titleLocArgs = notification.titleLocArgs,
        androidChannelId = notification.androidChannelId,
        tag = notification.tag,
        clickAction = notification.clickAction,
        emailEnabled = notification.emailEnabled == true,
        emailTitle = notification.emailTitle,
        emailBody = notification.emailBody,
        mutableContent = notification.mutableContent == true,
        createdAt = notification.createdAt?.toInstant(),
        updatedAt = notification.updatedAt?.toInstant()
    )

    override fun equals(other: Any?): Boolean = equalTo(
        other,
        FcmNotificationDto::delivered,
        FcmNotificationDto::ttlSeconds,
        FcmNotificationDto::scheduledTime,
        FcmNotificationDto::title,
        FcmNotificationDto::body,
        FcmNotificationDto::type,
        FcmNotificationDto::appPackage,
        FcmNotificationDto::sourceType,
    )

    override fun hashCode(): Int {
        return Objects.hash(
            scheduledTime,
            delivered,
            title,
            body,
            ttlSeconds,
            type,
            appPackage,
            sourceType,
        )
    }

    override fun toString(): String {
        return "FcmNotificationDto(id=$id, scheduledTime=$scheduledTime, delivered=$delivered, title=$title, body=$body, ttlSeconds=$ttlSeconds, sourceId=$sourceId, fcmMessageId=$fcmMessageId, fcmTopic=$fcmTopic, fcmCondition=$fcmCondition, type=$type, appPackage=$appPackage, sourceType=$sourceType, additionalData=$additionalData, priority=$priority, sound=$sound, badge=$badge, subtitle=$subtitle, icon=$icon, color=$color, bodyLocKey=$bodyLocKey, bodyLocArgs=$bodyLocArgs, titleLocKey=$titleLocKey, titleLocArgs=$titleLocArgs, androidChannelId=$androidChannelId, tag=$tag, clickAction=$clickAction, emailEnabled=$emailEnabled, emailTitle=$emailTitle, emailBody=$emailBody, mutableContent=$mutableContent, createdAt=$createdAt, updatedAt=$updatedAt)"
    }
}
